In [101]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [102]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [103]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [104]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [105]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
scaler = StandardScaler()
df_market_data_scaled = scaler.fit_transform(df_market_data)


In [106]:
# Create a DataFrame with the scaled data
df_market_data_scaled = pd.DataFrame(df_market_data_scaled, columns=df_market_data.columns)

# Copy the crypto names from the original data
df_market_data_scaled['coin_id']=df_market_data.index
# Set the coinid column as index
df_market_data_scaled.set_index('coin_id',inplace=True)

# Display sample data
df_market_data_scaled.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

---

### Find the Best Value for k Using the Original Data.

In [107]:
# Create a list with the number of k-values from 1 to 11
k_values=list(range(1,12))

In [108]:
# Create an empty list to store the inertia values
inertia=[]

# Create a for loop to compute the inertia with each possible value of k
for k in k_values:
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
    km = KMeans(n_clusters=k)
# 2. Fit the model to the data using `df_market_data_scaled`
    km.fit(df_market_data_scaled )
# 3. Append the model.inertia_ to the inertia list
    inertia.append(km.inertia_)

c:\Users\mcael\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:1036: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


In [109]:
print(inertia)

[287.0, 195.82021818036043, 123.19048183836958, 79.02243535120975, 65.30237914162498, 52.83658827713281, 43.91469044757747, 36.791069782922925, 32.62143314033297, 29.087571155722728, 25.49716247567548]


In [110]:
# Create a dictionary with the data to plot the Elbow curve
dict_elbow_curve = {"k_values": k_values, "inertia": inertia}

# Create a DataFrame with the data to plot the Elbow curve
df_elbow_curve = pd.DataFrame(dict_elbow_curve)

In [111]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
df_elbow_curve.hvplot.line(x="k_values", y="inertia", title="Elbow Curve", xticks=k_values)


:Curve   [k_values]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** 4

---

### Cluster Cryptocurrencies with K-means Using the Original Data

In [112]:
# Initialize the K-Means model using the best value for k
km = KMeans(n_clusters=4)

In [113]:
# Fit the K-Means model using the scaled data
km.fit(df_market_data_scaled)

KMeans(n_clusters=4)

In [114]:
# Predict the clusters to group the cryptocurrencies using the scaled data
labels = km.predict(df_market_data_scaled)

# Print the resulting array of cluster values.
print(labels)

# Print the cluster centers
cluster_centers = km.cluster_centers_
print(cluster_centers)


[3 3 0 0 3 3 3 3 3 0 0 0 0 3 0 3 0 0 3 0 0 3 0 0 0 0 0 0 3 0 0 0 1 3 0 0 2
 0 0 0 0]
[[ 0.03258562 -0.56998841 -0.4827023  -0.25453208 -0.07913767 -0.18795807
  -0.18203893]
 [-4.98104189 -0.04517829 -1.20695612 -1.21212587  0.04773554  4.63238025
   6.08862545]
 [ 1.04553034 -0.61832816  2.90705393  5.35145461  4.76991278  3.14887546
   1.34848839]
 [ 0.23756041  1.19101578  0.83462785  0.19065425 -0.212313   -0.22264199
  -0.20800783]]


In [115]:
# Create a copy of the DataFrame
df_market_data_scaled_cpy=df_market_data_scaled.copy()

In [116]:
# Add a new column to the DataFrame with the predicted clusters
df_market_data_scaled_cpy['clusters']=labels


# Display sample data
df_market_data_scaled_cpy.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  clusters  
coin_id                                             
bitcoin                        -0.251637         3  
ethereum                       -0.199352         3  
tether                         -0.282061         0  
ripple                         -0.295546         0  
bitcoin-cash                   -0.270317         3

In [117]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
x="price_change_percentage_24h"
y="price_change_percentage_7d"
df_market_data_scaled_cpy.reset_index(inplace=True)
scatter=df_market_data_scaled_cpy.hvplot.scatter(x, y, c="clusters", hover_cols=["coin_id"],title="Scatter Plot of cryptocurrency clusters")# * cluster_centers.hvplot.scatter(x,y, color='red', marker='x', size=200)
scatter

:Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,clusters,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [118]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components=3)

In [119]:
# Use the PCA model with `fit_transform` to reduce to 
# three principal components.
market_data_scaled_pca=pca.fit_transform(df_market_data_scaled)
# View the first five rows of the DataFrame. 
df_market_data_scaled_pca=pd.DataFrame(market_data_scaled_pca,columns=["PC1","PC2","PC3"])
df_market_data_scaled_pca.head()

PC1       PC2       PC3
0 -0.600667  0.842760  0.461595
1 -0.458261  0.458466  0.952877
2 -0.433070 -0.168126 -0.641752
3 -0.471835 -0.222660 -0.479053
4 -1.157800  2.041209  1.859715

In [120]:
# Retrieve the explained variance to determine how much information 
# can be attributed to each principal component.
exp_var_pca=pca.explained_variance_ratio_
print(exp_var_pca)

[0.3719856  0.34700813 0.17603793]


In [121]:
import numpy as np
np.sum(pca.explained_variance_ratio_[:3])

0.895031657030984

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** 0.895031657030984

In [122]:
# Create a new DataFrame with the PCA data.


# Creating a DataFrame with the PCA data
df_market_data_scaled_pca=pd.DataFrame(market_data_scaled_pca,columns=["PC1","PC2","PC3"])
# Copy the crypto names from the original data
coin_id=df_market_data.index

# Set the coinid column as index
df_market_data_scaled_pca.set_index(coin_id,inplace=True)

# Display sample data
df_market_data_scaled_pca.head()

PC1       PC2       PC3
coin_id                                   
bitcoin      -0.600667  0.842760  0.461595
ethereum     -0.458261  0.458466  0.952877
tether       -0.433070 -0.168126 -0.641752
ripple       -0.471835 -0.222660 -0.479053
bitcoin-cash -1.157800  2.041209  1.859715

---

### Find the Best Value for k Using the PCA Data

In [123]:
# Create a list with the number of k-values from 1 to 11
k_values_pca=list(range(1,12))

In [124]:
# Create an empy list to store the inertia values
inertia_pca=[]

# Create a for loop to compute the inertia with each possible value of k
for k in k_values_pca:
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
    km_pca = KMeans(n_clusters=k)
# 2. Fit the model to the data using `df_market_data_pca`
    km_pca.fit(df_market_data_scaled_pca)
# 3. Append the model.inertia_ to the inertia list
    inertia_pca.append(km_pca.inertia_)

c:\Users\mcael\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:1036: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


In [125]:
# Create a dictionary with the data to plot the Elbow curve
dict_elbow_curve_pca = {"k_values": k_values_pca, "inertia": inertia_pca}
# Create a DataFrame with the data to plot the Elbow curve
df_elbow_curve_pca = pd.DataFrame(data=dict_elbow_curve_pca)


In [126]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.

df_elbow_curve_pca.hvplot.line(x="k_values", y="inertia", title="Elbow Curve using PCA", xticks=k_values_pca)


:Curve   [k_values]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** 4


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** No, it's the same value

### Cluster Cryptocurrencies with K-means Using the PCA Data

In [127]:
# Initialize the K-Means model using the best value for k
km_pca = KMeans(n_clusters=4)

In [128]:
# Fit the K-Means model using the PCA data
km_pca.fit(df_market_data_scaled_pca)

KMeans(n_clusters=4)

In [129]:
# Predict the clusters to group the cryptocurrencies using the PCA data
labels_pca = km_pca.predict(df_market_data_scaled_pca)
# Print the resulting array of cluster values.
print(labels_pca)
# Predict the clusters to group the cryptocurrencies using the scaled data
cluster_centers_pca = km_pca.cluster_centers_
#print(cluster_centers_pca)


[3 3 1 1 3 3 3 3 3 1 1 1 1 3 1 3 1 1 3 1 1 3 1 1 1 1 1 1 3 1 1 1 2 3 1 1 0
 1 1 1 1]


In [130]:
# Create a copy of the DataFrame with the PCA data
df_market_data_scaled_pca_cpy=df_market_data_scaled_pca.copy()

# Add a new column to the DataFrame with the predicted clusters
df_market_data_scaled_pca_cpy['clusters_pca']=labels_pca

# Display sample data
df_market_data_scaled_pca_cpy.head()


PC1       PC2       PC3  clusters_pca
coin_id                                                 
bitcoin      -0.600667  0.842760  0.461595             3
ethereum     -0.458261  0.458466  0.952877             3
tether       -0.433070 -0.168126 -0.641752             1
ripple       -0.471835 -0.222660 -0.479053             1
bitcoin-cash -1.157800  2.041209  1.859715             3

In [131]:
x="PC1"
y="PC2"
df_market_data_scaled_pca_cpy.reset_index(inplace=True)
scatter_pca = df_market_data_scaled_pca_cpy.hvplot.scatter(x, y, c="clusters_pca", hover_cols=["coin_id"],title="Scatter Plot of cryptocurrency clusters using PCA")
scatter_pca

:Scatter   [PC1]   (PC2,clusters_pca,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [132]:
# Composite plot to contrast the Elbow curves

p = df_elbow_curve.hvplot.line(x="k_values", y="inertia", title="Elbow Curve without using PCA", xticks=k_values,color='red') #df_elbow_curve_pca.hvplot.line(x="k_values", y="inertia", title="Elbow Curve using PCA", xticks=k_values_pca)
p_pca = df_elbow_curve_pca.hvplot.line(x="k_values", y="inertia", title="Elbow Curve using PCA", xticks=k_values_pca,color='blue')
composite_plot = (p * p_pca).opts(title='Elbow Curves with (blue) and without PCA (red)', xlabel='k_values', ylabel='inertia')

# Display the composite plot
composite_plot

:Overlay
   .Curve.I  :Curve   [k_values]   (inertia)
   .Curve.II :Curve   [k_values]   (inertia)

In [133]:
# Composite plot to contrast the clusters
# YOUR CODE HERE!
composite_scatter = (scatter + scatter_pca).cols(2)
composite_scatter

:Layout
   .Scatter.I  :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,clusters,coin_id)
   .Scatter.II :Scatter   [PC1]   (PC2,clusters_pca,coin_id)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** We can see that K-means using PCA shows a better separation of clusters, which means that the principal components captures the structure of the data better.